### ISSUE TITLE
Rustix recompiles every time switching between Intellij and CLI

### ISSUE THREAD

**Issue by howardjohn** [bytecodealliance/rustix#562]
This is maybe not even a rustix bug, but may help someone else seeing the same issue.

Rustix's build.rs has `cargo:rerun-if-env-changed=RUSTC`.

When I run `cargo check` from my terminal, I see `RUSTC=rustc`. When I run it from intellij, I get `RUSTC=$HOME/.cargo/bin/rustc`. This means each time the build is invalidated.

For now I am just doing `export RUSTC=$HOME/.cargo/bin/rustc` to work around this

**Comment by sunfishcode** [comment]
Rustix needs to autodetect features from the rustc, in particular to know whether it can used `OwnedFd` from std or whether it needs to use its own copy. If it's possible that `cargo check` in your terminal could use a different version of rustc than what intellij uses, then rustix does need to rerun its build script.

Another consideration here is that https://github.com/intellij-rust/intellij-rust/issues/10186 is a report where rustix may not be running its build script in a situation where it should. It's not yet clear what the problem is.

Also relevant in this space is that my earlier fix had a bug, which I'm now fixing in https://github.com/bytecodealliance/rustix/pull/563.

**Comment by sunfishcode** [comment]
In Rust 1.68, `cargo build --verbose` now includes a `Dirty` message the indicates why it triggered a rebuild, which would be useful to see when this happens.

**Comment by Nutomic** [comment]
Im have the same problem described in this issue, this is the output for rustix from `cargo check --verbose`:
```
Running `/usr/bin/sccache rustc --crate-name rustix --edition=2018 /home/felix/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rustix-0.36.9/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=153 --crate-type lib --emit=dep-info,metadata -C embed-bitcode=no -C debuginfo=0 --cfg 'feature="default"' --cfg 'feature="fs"' --cfg 'feature="io-lifetimes"' --cfg 'feature="libc"' --cfg 'feature="mm"' --cfg 'feature="std"' --cfg 'feature="use-libc-auxv"' -C metadata=fe6d77cbb4767f55 -C extra-filename=-fe6d77cbb4767f55 --out-dir /home/felix/workspace/webb/relayer/target/debug/deps -C strip=symbols -L dependency=/home/felix/workspace/webb/relayer/target/debug/deps --extern bitflags=/home/felix/workspace/webb/relayer/target/debug/deps/libbitflags-d53a0f852625ca18.rmeta --extern io_lifetimes=/home/felix/workspace/webb/relayer/target/debug/deps/libio_lifetimes-1327a00c9e419885.rmeta --extern libc=/home/felix/workspace/webb/relayer/target/debug/deps/liblibc-6ce3c9449f468f7a.rmeta --extern linux_raw_sys=/home/felix/workspace/webb/relayer/target/debug/deps/liblinux_raw_sys-d6534f431c32c99e.rmeta --cap-lints allow --cfg rustc_attrs --cfg linux_raw --cfg core_intrinsics --cfg asm`
Running `/usr/bin/sccache rustc --crate-name rustix --edition=2018 /home/felix/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rustix-0.36.9/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=153 --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no --cfg 'feature="default"' --cfg 'feature="fs"' --cfg 'feature="io-lifetimes"' --cfg 'feature="libc"' --cfg 'feature="mm"' --cfg 'feature="std"' --cfg 'feature="use-libc-auxv"' -C metadata=2b984a9b6c932294 -C extra-filename=-2b984a9b6c932294 --out-dir /home/felix/workspace/webb/relayer/target/debug/deps -C strip=symbols -L dependency=/home/felix/workspace/webb/relayer/target/debug/deps --extern bitflags=/home/felix/workspace/webb/relayer/target/debug/deps/libbitflags-c672ab23fb67ebd2.rmeta --extern io_lifetimes=/home/felix/workspace/webb/relayer/target/debug/deps/libio_lifetimes-c416a3d2a8a68533.rmeta --extern libc=/home/felix/workspace/webb/relayer/target/debug/deps/liblibc-f5cd561e13a0f6ad.rmeta --extern linux_raw_sys=/home/felix/workspace/webb/relayer/target/debug/deps/liblinux_raw_sys-bbd9f11733e8abe5.rmeta --cap-lints allow --cfg rustc_attrs --cfg linux_raw --cfg core_intrinsics --cfg asm`
```

**Comment by sunfishcode** [comment]
@Nutomic Are you using Rust 1.68, which has the new `Dirty` output?

**Comment by OvermindDL1** [comment]
I am on:
```
❯ rustc --version
rustc 1.68.2 (9eb3afe9e 2023-03-27)

❯ cargo --version
cargo 1.68.2 (6feb7c9cf 2023-03-26)
```
And I'm constantly having this issue as well.  Running `cargo build --verbose` and I see this `Dirty` message
```
Dirty rustix v0.37.4: the env variable RUSTC changed
```
Rustix version is 0.37.4 (required by `is-terminal` 0.4.6, which is required by `anstream` 0.2.6, which is required by `clap_builder` 4.2.0, which is required by `clap` 4.2.0, which is required by my projects).

Every time I `cargo build` it's having to recompile rustix.

**Comment by OvermindDL1** [comment]
Just for completion, here's the output of a `cargo build --verbose` that contains `rustix` on each line:
```
       Dirty rustix v0.37.4: the env variable RUSTC changed
   Compiling rustix v0.37.4
     Running `/home/overminddl1/rust/testing_constant_recompiling/target/debug/build/rustix-b26a51cba84e394d/build-script-build`
     Running `rustc --crate-name rustix --edition=2018 /home/overminddl1/.cargo/registry/src/github.com-1ecc6299db9ec823/rustix-0.37.4/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=132 --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg 'feature="default"' --cfg 'feature="io-lifetimes"' --cfg 'feature="libc"' --cfg 'feature="std"' --cfg 'feature="termios"' --cfg 'feature="use-libc-auxv"' -C metadata=60d941d26896765e -C extra-filename=-60d941d26896765e --out-dir /home/overminddl1/rust/testing_constant_recompiling/target/debug/deps -L dependency=/home/overminddl1/rust/testing_constant_recompiling/target/debug/deps --extern bitflags=/home/overminddl1/rust/testing_constant_recompiling/target/debug/deps/libbitflags-f255a966af175049.rmeta --extern io_lifetimes=/home/overminddl1/rust/testing_constant_recompiling/target/debug/deps/libio_lifetimes-0861c285f6a9ecc3.rmeta --extern libc=/home/overminddl1/rust/testing_constant_recompiling/target/debug/deps/liblibc-9de7ca31dbbda4df.rmeta --extern linux_raw_sys=/home/overminddl1/rust/testing_constant_recompiling/target/debug/deps/liblinux_raw_sys-10a54ce77ae265d4.rmeta --cap-lints allow --cfg linux_raw --cfg asm --cfg linux_like`
```

**Comment by OvermindDL1** [comment]
Also note, this happens between intellij and cli as well as rust-analyzer (via emacs) and cli.

**Comment by OvermindDL1** [comment]
Just testing an older version of rustix and it didn't have this issue, but rather if it was `cargo clean`'d first and then built in intellij then it would give errors about using nightly features on stable, yet worked when it was built in the cli first, which is odd as I don't even have the nightly compiler installed at all...

**Comment by OvermindDL1** [comment]
Just to remark on its impact, it's adding about 12.6s of time to running the tests (of which they themselves take less than 2 seconds), and this is every time they are run (which I do on-the-fly via cargo-watch), it's quite an added delay compared to the usual compiling time of just my code of less than 2 seconds to compile as well when rustix isn't recompiling.

**Comment by sunfishcode** [comment]
I've now backported the #576 fix to the 0.36 branch, and made a 0.36.14 release with it. It's also in 0.37.5 and later 0.37 releases.

**Comment by sunfishcode** [comment]
#576 seems to have fixed this, as I haven't heard of any further reports of this problem since then. If anyone does see this happen again, please file a new issue!
